### DISCUSSION TITLE
New update in GFW

### DISCUSSION THREAD

**Discussion by itsamirhn** [v2fly/v2ray-core#2102] [General]
As you may knew, the Great Firewall rolled out an update that blocks TLS-based circumvention tools, such as trojan, Xray, V2Ray...
More detail;
https://www.top10vpn.com/best-vpn/china/

I was wondering how we are going to bypass censorship if GFW completely blocks this service?

Is there any solution to avoid this blocking?

Or other than v2ray, Is there any undetectable protocol that we can use?

**Comment by kingwilliam** [comment]
please calm down, until now I still using v2fly for daily and work perfectly.

My config:
```
1st client(v2fly vless+tls+grpc) server(haproxy+v2fly) self sign certificate
2nd client(gost tls+grpc) server(haproxy+gost) self sign certificate
final backup client(gost ss+tcp+shadow-tls) server(haproxy+shadow-tls+gost)
```

ALL still work fine

**Reply by itsamirhn** [reply]
Would you please give a tutorial for setting up your "final backup" solution?
It's ok to be in Chinese language.

**Comment by kingwilliam** [comment]
建议先了解 [shadow-tls](https://github.com/ihciah/shadow-tls) 是什麽和如何工作

concept
```
client
gost/v2 -(ss+tcp)-> shadow-tls

server
shadow-tls -> gost/v2
```


client
```
gost -L :8080 -F ss://AEAD_CHACHA20_POLY1305:123456@127.0.0.1:8001

shadowtls client --listen :8001 --password stls-456789 --server {IP/Domain}:443 --sni {伪装domain}
```


server 
```
shadowtls server --listen :443 --password stls-456789 --server 127.0.0.1:8002 --tls {伪装domain}:443

gost -L ss://AEAD_CHACHA20_POLY1305:123456@127.0.0.1:8002
```

gost/v2 也可以连接shadow-tls

关於shadow-tls问题, 可看看以下连结
[报告: 中间盒如何识别代理流量](https://telegra.ph/%E6%8A%A5%E5%91%8A-%E4%B8%AD%E9%97%B4%E7%9B%92%E5%A6%82%E4%BD%95%E8%AF%86%E5%88%AB%E4%BB%A3%E7%90%86%E6%B5%81%E9%87%8F-10-17)


补充:
以上方式也不是绝对安全, 只可算是最后一根救命草. 在 墙 下,没有一个科学上网是绝对安全和不能被探测. 我们一般网民只能尽量避免给 墙 钉上. 至于如何不被钉上 只能八仙过海各显神通.

最後
顺道说句, 什麽是 精准识别? openvpn 就是精准识别的好例子.
今天的情况 只算是 提高识别率而已

**Reply by seakfind** [reply]
@kingwilliam Would you mind sharing the `haproxy` configuration?

**Reply by cxjava** [reply]
这个客户端和服务端的配置，用go-gost/gost 和 ginuerzh/gost ，shadow-tls 0.2.8 都不行。不知道楼主怎么成功的

**Comment by kingwilliam** [comment]
I highly recommend visit [XTLS Vision, TLS in TLS, to the star and beyond #1295](https://github.com/XTLS/Xray-core/discussions/1295) first. To know what is "TLS over TLS" in Today, and why grpc can SURVIVE at this moment(just only more a little bit).

remarks: 
0. *** ONLY FOR REFERENCE ***
1. i use ..... to hidden common command
2. in "frontend in-path_filter" ca-file CA_crt.pem verify optional, need client cert to access v2_gost server.
- 2a if without client cert, only to default-web-server
- 2b so, mobile may be not accepted on this config

```
global
        .....
        .....
        .....
        maxconn 5120
        tune.ssl.default-dh-param 2048
        tune.h2.initial-window-size 10485760

        ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        ssl-default-bind-options prefer-client-ciphers no-sslv3 no-tlsv10 no-tlsv11 no-tlsv12 no-tls-tickets

        ssl-default-server-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tlsv12 no-tls-tickets


defaults
        log     global
        mode    tcp
        option  tcplog
        option  tcp-smart-accept
        option  tcp-smart-connect
        option  dontlognull
        http-reuse always
        timeout http-request 5s
        timeout connect 5s
        timeout client  5m
        timeout server  5m
        timeout tunnel  1h
        .....
        .....
        .....


frontend in-00
    bind *:443 alpn h2 allow-0rtt
    acl v2_gost_tls    req.ssl_sni -i www.v2_gost.com
    acl shadow_tls     req.ssl_sni -i www.microsoft.com

    tcp-request inspect-delay 5s
    tcp-request content reject if !{ req_ssl_hello_type 1 }
    tcp-request content reject if !v2_gost_tls !shadow_tls

    use_backend bkend_v2_gost_tls     if v2_gost_tls
    use_backend bkend_shadow_tls      if shadow_tls



backend bkend_v2_gost_tls
    server svr_v2_gost abns@path_filter.sock send-proxy-v2

backend bkend_shadow_tls
    server svr_shadow 127.0.0.1:8001




frontend in-path_filter
    bind abns@path_filter.sock ssl crt www.v2_gost.com.pem ca-file CA_crt.pem verify optional accept-proxy alpn h2 allow-0rtt

    acl chk-02_h2 ssl_fc_alpn -i h2
    acl chk-02_ccrt  ssl_fc_has_crt

    acl grpc-01a path  -m str /gost-gRPC/tunnel
    acl grpc-01b path  -m str /v2-vl-gRPC/Tun

    use_backend bkend_go-h2-01    if grpc-01a chk-02_h2 chk-02_ccrt
    use_backend bkend_v2-grpc-01  if grpc-01b chk-02_h2 chk-02_ccrt

    default_backend default-web-server


backend bkend_go-h2-01
    mode http
    server svr_go-h2-01 127.0.0.1:11112 proto h2


backend bkend_v2-grpc-01
    mode http
    server svr_v2_h2-01 127.0.0.1:11102 proto h2





backend default-web-server
    mode http
    server svr_local-99 127.0.0.1:8000
```

**Reply by seakfind** [reply]
Very good. Thanks!

**Comment by jpalidev** [comment]
You can also combine Shadow-TLS with V2Ray, like this:

https://palitechsociety.blogspot.com/2022/11/vmess-shadow-tls.html

Shadow-TLS is relatively new but has already come to the attention of the firewall-makers:

https://www.viewintech.com/html/articledetails.html?newsId=23

Despite this, an update to the blog dated 2022-11-13 says that Shadow-TLS is still working:

https://www.ihcblog.com/a-better-tls-obfs-proxy
